pkg/util/kubernetes/kubelet/kubelet.go

```diff
@@ -632,74 +630,96 @@ func potentialKubeletHostsFilter(h *host) {
 
 	hostnamesKeys := make(map[string]bool)
 	hostnames := []string{}
-	for _, hostname := range h.hostnames {
+	for _, hostname := range hosts.hostnames {
 		if _, check := hostnamesKeys[hostname]; !check {
 			hostnamesKeys[hostname] = true
 			hostnames = append(hostnames, hostname)
 		}
 	}
 
-	h.ips = ips
-	h.hostnames = hostnames
+	hosts.ips = ips
+	hosts.hostnames = hostnames
 }
 
-func (ku *KubeUtil) setKubeletHost(h *host, httpsPort, httpPort int) error {
-	var err error
-	for _, ip := range h.ips {
-		log.Infof("Trying to use host %s with HTTPS", ip)
-		ku.kubeletHost = ip
-		err := ku.setupKubeletApiClient()
-		if err != nil {
-			return err
-		}
+// setKubeletHost select a kubelet host from potential kubelet hosts
+// the method check HTTPS connections first and prioritize ips over hostnames
+func (ku *KubeUtil) setKubeletHost(hosts *connectionInfo, httpsPort, httpPort int) error {
+	var connectionErrors []error
+	kubeletHost, errors := selectFromPotentialHostsHTTPS(ku, hosts.ips, httpsPort)
+	if kubeletHost != "" && errors == nil {
+		log.Infof("%s is set as kubelet host", kubeletHost)
+		return nil
+	} else {
+		connectionErrors = append(connectionErrors, errors...)
+	}
 
-		err = checkKubeletHTTPSConnection(ku, httpsPort)
-		if err == nil {
-			log.Infof("Can connect to kubelet using %s and HTTPS, setting up %s as kubelet host", ip, ip)
-			ku.kubeletHost = ip
-			return nil
-		}
+	kubeletHost, errors = selectFromPotentialHostsHTTPS(ku, hosts.hostnames, httpsPort)
+	if kubeletHost != "" && errors == nil {
+		log.Infof("%s is set as kubelet host", kubeletHost)
+		return nil
+	} else {
+		connectionErrors = append(connectionErrors, errors...)
+	}
+
+	kubeletHost, errors = selectFromPotentialHostsHTTP(hosts.ips, httpPort)
+	if kubeletHost != "" && errors == nil {
+		ku.kubeletHost = kubeletHost
+		log.Infof("%s is set as kubelet host", kubeletHost)
+		return nil
+	} else {
+		connectionErrors = append(connectionErrors, errors...)
+	}
+
+	kubeletHost, errors = selectFromPotentialHostsHTTP(hosts.hostnames, httpPort)
+	if kubeletHost != "" && errors == nil {
+		ku.kubeletHost = kubeletHost
+		log.Infof("%s is set as kubelet host", kubeletHost)
+		return nil
+	} else {
+		connectionErrors = append(connectionErrors, errors...)
 	}
 
-	for _, hostname := range h.hostnames {
-		log.Infof("Trying to use host %s with HTTPS", hostname)
-		ku.kubeletHost = hostname
+	log.Errorf("Cannot set a valid kubeletHost: cannot connect to kubelet using any of the given hosts: %v %v, Errors: %v", hosts.ips, hosts.hostnames, connectionErrors)
```

Comment:
How does this look in a real case usage? I'm worried this will look bad in the agent logs (many errors in a row, unstructured, not all of errors being relevant since we probably only care about the last ones...)

Reply:
I agree, logs are not well structured this way although I think we care about all errors not only the last ones as in unit and integration tests we check HTTPS connection errors (especially errors related to TLS settings) which are the first errors we get in our logic. That's why I'm collecting all connection errors and returning them in the error string in the next line in the function.

Regarding logs:

- I can modify `connectionInfo` and make it like this to track errors related to each host (will also dedupe hosts automatically using maps)
```
type connectionInfo struct {
	ips       map[string]error
	hostnames map[string]error
}
```
- otherwise, I suggest a simpler solution by logging every error as Warn when checking connection and appendig the error in `selectFromPotentialHosts` functions, not in `setKubeletHost`. This will be structured and more meaningful in agent logs comparing to logging all the list of errors in `setKubeletHost`.
WDYT?

Reply:
I like both options. To me the most important part is that it's clear in the error message we log that the first errors are not blocking and that the agent is gonna try other things before failing. Otherwise we'll get support tickets about these error messages when they're actually not a problem. Logging them as `warn` is a good first step 👍 maybe wrapping them in `info` logs that say `Trying several connection methods to locate the kubelet...` and `Connection to the kubelet succeeded!` or `All connection attempts to the Kubelet failed.` would help as well.